### PR TITLE
Create stub of server with express and socket.io

### PR DESCRIPTION
**Pull request by asross**
This just adds a server with express.js and socket.io, plus a hello world page that connects to them.

---
This change is [Reviewable](https://reviewable.io/reviews/pair-code/federated-learning/2)

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "compilerOptions": {
-    "module": "ES2015",
+    "module": "commonjs",
```
Comment (asross) [inline tsconfig.json]: I noticed that in `tfjs`, there are two different `tsconfig` files; one which uses `ES2015` and ends up in `dist-es6`, and another which uses es5-compatible `commonjs` and ends up in `dist`. For our application, does it make sense to just use a single `tsconfig` with commonjs and writing into dist?

**Comment by dsmilkov** [comment]
:lgtm_strong: Nice ! LGTM. Left comment of small comments, but nothing that needs me to explicitly revisit, so feel free to submit after.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[package.json, line 12 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEKz_xa4oCSpx6rPNhr:-LEKz_xb-Fx6tv9rDZfe:b7ty64g) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/package.json#L12)):*
> ```JSON
>   "license": "Apache-2.0",
>   "scripts": {
>     "start": "node dist/server/server.js",
> ```

tip: use `ts-node` (it's like node but you can pass typescript files and the compilation happens transparently). We use it in tfjs-node. See [here](https://github.com/tensorflow/tfjs-node/blob/master/package.json#L19)

---

*[tsconfig.json, line 3 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEKnRuNedZb3-aKXlzB-r1-3:-LEKzreh59wc0GC6YL59:b-qglaz0) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/tsconfig.json#L3)):*
<details><summary><i>Previously, asross (Andrew Ross) wrote…</i></summary><blockquote>

I noticed that in `tfjs`, there are two different `tsconfig` files; one which uses `ES2015` and ends up in `dist-es6`, and another which uses es5-compatible `commonjs` and ends up in `dist`. For our application, does it make sense to just use a single `tsconfig` with commonjs and writing into dist?
</blockquote></details>

makes sense. we can revisit later but this is good for now!

---

*[views/index.html, line 2 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEL-7QH8Rnv4yDCxTp-:-LEL-7QH8Rnv4yDCxTp0:bdzozop) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/views/index.html#L2)):*
> ```HTML
> <html>
>   <head>
> ```

it might make sense to move this `view/index.html` under `src/client`. what do you think?

---


*Comments from [Reviewable](https://reviewable.io/reviews/pair-code/federated-learning/2#-:-LEL-HaYBIR5r8RuVtRE:b3w8uck)*

**Comment by nsthorat** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[src/server/server.ts, line 24 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEL-c5z8lu9eMatFOEn:-LEL-c5z8lu9eMatFOEo:b-swa7qd) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/src/server/server.ts#L24)):*
> ```XML
> import * as socketIO from 'socket.io';
> import * as path from 'path';
> import { Request, Response } from 'express';
> ```

Are you using vscode? if so, you should install clang-format and tslint extensions.

---

*[views/index.html, line 5 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEL-mdO5yF-GH8we_Oz:-LEL-mdO5yF-GH8we_P-:bjw9eg7) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/views/index.html#L5)):*
> ```HTML
>     <script src="/socket.io/socket.io.js"></script>
>     <script>
>       var socket = io.connect('http://localhost:3000');
> ```

prefer const / let over var (in this case use const).

---


*Comments from [Reviewable](https://reviewable.io/reviews/pair-code/federated-learning/2)*

**Comment by dsmilkov** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[views/index.html, line 2 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEL-7QH8Rnv4yDCxTp-:-LEL06tt6gX9FaYQX79p:b-p84wx8) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/views/index.html#L2)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

it might make sense to move this `view/index.html` under `src/client`. what do you think?
</blockquote></details>

actually, this is more of a demo (user of the library) so might make sense to rename to demo/index.html and keep it top-level. `src/client/...` should be reserved for reusable client code that any demo can use.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pair-code/federated-learning/2)*

**Comment by asross** [comment]
*[src/server/server.ts, line 24 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEL-c5z8lu9eMatFOEn:-LEL5qnFBGwA-qkqCFnv:bo211pr) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/src/server/server.ts#L24)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Are you using vscode? if so, you should install clang-format and tslint extensions.
</blockquote></details>

Sorry, didn't have them installed on my local machine. This should be fixed now.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pair-code/federated-learning/2)*

**Comment by asross** [comment]
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[package.json, line 12 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEKz_xa4oCSpx6rPNhr:-LEL60FqD3pqvKcIcC2c:b-745hf2) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/package.json#L12)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

tip: use `ts-node` (it's like node but you can pass typescript files and the compilation happens transparently). We use it in tfjs-node. See [here](https://github.com/tensorflow/tfjs-node/blob/master/package.json#L19)
</blockquote></details>

Hmm... I just tried this, but it seems to run the resulting compiled file from its current location (in `src`, not in `dist`). This breaks the path to `index.html`.

Would everything work if we moved `index.html` under `src/demo/` and made `dist/` a perfect mirror of the directory structure of `src/`?

---

*[views/index.html, line 2 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEL-7QH8Rnv4yDCxTp-:-LEL5meiDSvZ6a59TcJg:bj86og5) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/views/index.html#L2)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

actually, this is more of a demo (user of the library) so might make sense to rename to demo/index.html and keep it top-level. `src/client/...` should be reserved for reusable client code that any demo can use.
</blockquote></details>

K, done!

---


*Comments from [Reviewable](https://reviewable.io/reviews/pair-code/federated-learning/2)*

**Comment by dsmilkov** [comment]
:lgtm_strong: (feel free to submit whenever you get LGTM - no need to wait for another pass)

---

Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/pair-code/federated-learning/2#-:-LEMDfAt0T80FFwVtp-Q:b2aqqyd)*

File: tsconfig.json (hunk 1 above)
Comment (dsmilkov) [reply tsconfig.json]: makes sense. we can revisit later but this is good for now!

**Comment by dsmilkov** [comment]
Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[package.json, line 12 at r1](https://reviewable.io/reviews/pair-code/federated-learning/2#-LEKz_xa4oCSpx6rPNhr:-LEMchMqCSqlCB09Ic-G:b-kjgty3) ([raw file](https://github.com/pair-code/federated-learning/blob/0142dd9438d335b3cbf287bde558312ac35f42aa/package.json#L12)):*
<details><summary><i>Previously, asross (Andrew Ross) wrote…</i></summary><blockquote>

Hmm... I just tried this, but it seems to run the resulting compiled file from its current location (in `src`, not in `dist`). This breaks the path to `index.html`.

Would everything work if we moved `index.html` under `src/demo/` and made `dist/` a perfect mirror of the directory structure of `src/`?
</blockquote></details>

I see what you mean. I was thinking that the demo (e.g. a game) will have it's own static resources and its own http server, and the server.js would only be responsible for communicating weights, but not serving any static resources.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pair-code/federated-learning/2)*

**Pull request merged**